File: MaxHeap.py
import random
import time

class MaxHeap:
    def __init__(self):
        self.baum = []
        self.swap_count = 0

    def fill_with_random(self, n, *, lo=0, hi=100):
        self.baum = [random.randrange(lo, hi) for _ in range(n)]
        self.swap_count = 0

    def _lchild(self, i): return 2*i + 1
    def _rchild(self, i): return 2*i + 2

    def _swap(self, i, j):
        self.baum[i], self.baum[j] = self.baum[j], self.baum[i]
        self.swap_count += 1

    def _heapify(self, i, size):
        while True:
            left  = self._lchild(i)
            right = self._rchild(i)
            largest = i
            if left < size and self.baum[left]  > self.baum[largest]:
                largest = left
            if right < size and self.baum[right] > self.baum[largest]:
                largest = right
            if largest != i:
                self._swap(i, largest)
                i = largest
            else:
                break

    def _build_heap(self):
        n = len(self.baum)
        for i in range(n//2 - 1, -1, -1):
            self._heapify(i, n)

    def heap_sort(self):
        self.swap_count = 0
        self._build_heap()
        n = len(self.baum)
        for end in range(n-1, 0, -1):
            self._swap(0, end)
            self._heapify(0, end)
        self.baum.reverse()

    def print_tree(self):
        def _rec(idx, depth):
            if idx >= len(self.baum):
                return
            _rec(self._rchild(idx), depth+1)
            print("   " * depth + str(self.baum[idx]))
            _rec(self._lchild(idx), depth+1)
        _rec(0, 0)

if __name__ == "__main__":
    heap = MaxHeap()
    heap.fill_with_random(10)
    print("Unsorted:", heap.baum)

    start = time.perf_counter()
    heap.heap_sort()
    duration = time.perf_counter() - start

    print("Sorted (desc):", heap.baum)
    print(f"Sortierdauer: {duration*1e3:.3f} ms ({duration*1e9:.0f} ns)")
    print("Anzahl Swaps:", heap.swap_count)
    print("\nHeap-Tree:")
    heap.print_tree()

File: MaxHeap.java
import java.util.ArrayList;
import java.util.Collections;

public class MaxHeap {
    private ArrayList<Integer> baum;
    private long swapCount;

    public MaxHeap() {
        baum = new ArrayList<>();
        swapCount = 0;
    }

    public void fillWithRandomNumbers(int i) {
        baum.clear();
        for (int j = 0; j < i; j++) {
            baum.add((int) (Math.random() * 1000));
        }
        swapCount = 0;
    }

    private int getLChild(int i) {
        return 2 * i + 1;
    }

    private int getRChild(int i) {
        return 2 * i + 2;
    }

    private void swap(int i, int j) {
        Collections.swap(baum, i, j);
        swapCount++;
    }

    private void heapify(int i, int r) {
        while (true) {
            int left  = getLChild(i);
            int right = getRChild(i);
            int largest = i;

            if (left < r && baum.get(left)  > baum.get(largest)) largest = left;
            if (right < r && baum.get(right) > baum.get(largest)) largest = right;

            if (largest != i) {
                swap(i, largest);
                i = largest;
            } else {
                break;
            }
        }
    }

    private void buildHeap() {
        int n = baum.size();
        for (int i = n/2 - 1; i >= 0; i--) {
            heapify(i, n);
        }
    }

    public void heapSort() {
        swapCount = 0;
        buildHeap();
        int n = baum.size();
        for (int r = n - 1; r > 0; r--) {
            swap(0, r);
            heapify(0, r);
        }
        // Einfachste Lösung: Umkehren der Liste
        Collections.reverse(baum); 
    }

    public long getSwapCount() {
        return swapCount;
    }

    public static void main(String[] args) {
        MaxHeap heap = new MaxHeap();
        heap.fillWithRandomNumbers(10);
        System.out.println("Unsorted: " + heap.baum);

        long start = System.nanoTime();
        heap.heapSort();
        long end = System.nanoTime();

        System.out.println("Sorted: " + heap.baum);
        System.out.printf("Sortierdauer: %d ns (%.3f ms)%n", end - start, (end - start)/1e6);
        System.out.println("Anzahl Swaps: " + heap.getSwapCount());
    }
}